bitcoin/src/blockdata/transaction.rs

```diff
@@ -693,7 +693,7 @@ impl Transaction {
 
     /// Returns the regular byte-wise consensus-serialized size of this transaction.
     #[inline]
-    pub fn size(&self) -> usize { self.scaled_size(1) }
+    pub fn size(&self) -> usize { self.scaled_size(1).to_wu() as usize }
```

Comment:
Can you justify why this cast is ok in a code comment please.

Reply:
This is surrounded by tests that can justify the change is ok.  For example, if this line is changed to:
 `pub fn size(&self) -> usize { (self.scaled_size(1).to_wu() + 1) as usize } `

then the following failures are revealed:

```
failures:                                                                                                                                                                                      
                                                                                                                                                                                               
---- blockdata::block::tests::block_test stdout ----                                                                                                                                           
thread 'blockdata::block::tests::block_test' panicked at 'assertion failed: `(left == right)`                                                                                                  
  left: `645`,                                                                                                                                                                                 
 right: `643`', bitcoin/src/blockdata/block.rs:493:9                                                                                                                                           
                                                                                                                                                                                               
---- blockdata::block::tests::segwit_block_test stdout ----                                                                                                                                    
thread 'blockdata::block::tests::segwit_block_test' panicked at 'assertion failed: `(left == right)`                                                                                           
  left: `4334`,                                                                                                                                                                                
 right: `4319`', bitcoin/src/blockdata/block.rs:535:9                                                                                                                                          
                                                                                                                                                                                               
---- blockdata::transaction::tests::nonsegwit_transaction stdout ----                                                                                                                          
thread 'blockdata::transaction::tests::nonsegwit_transaction' panicked at 'assertion failed: `(left == right)`                                                                                 
  left: `194`,                                                                                                                                                                                 
 right: `193`', bitcoin/src/blockdata/transaction.rs:1398:9                                                                                                                                    
                                                                                                                                                                                               
---- blockdata::transaction::tests::segwit_transaction stdout ----                                                                                                                             
thread 'blockdata::transaction::tests::segwit_transaction' panicked at 'assertion failed: `(left == right)`                                                                                    
  left: `194`,                                                                                                                                                                                 
 right: `193`', bitcoin/src/blockdata/transaction.rs:1440:9
```

Reply:
This PR does not change the code semantics.  Before the set of values returned from `scaled_size` was truncated to `usize`.  Now `scaled_size` returns a weight type which is not truncated depending on arch type.  However since `size()` still returns `usize`, a cast to `usize` is required.  I can base a PR off of this PR which changes the return type of `size()` as well.

Reply:
I'm OK ACKing even without an explanatory comment, but I would like to revisit this and maybe add some tests which attempt to get a `Weight` object to exceed 2^32 and see what happens. I believe both of you that (a) this cast is sketchy and (b) it matches the existing code.

Later on we should try to create an explicit internal invariant that we're never representing more than 2^31 wu using this type. Though no doubt we will have endless bikeshedding about the specific cap.

Having said that, I think the existing code is fine for basically all applications, which compute weights either for (a) things that the author's own code produced; and (b) things that came from the blockchain or p2p network. Neither of which can come close to overflowing 2^32.

Reply:
> I'm OK ACKing even without an explanatory comment, but I would like to revisit this and maybe add some tests which attempt to get a Weight object to exceed 2^32 and see what happens. 

If my understanding is right, I think what your asking would be hard to test since `usize` will be different on different arch type.  I put in a PR [here](https://github.com/rust-bitcoin/rust-bitcoin/pull/2047) to change the return type to `Weight` instead of `usize` which should prevent any ambiguity in arch types.  I'm not sure what else to test..